### PR TITLE
Remote config client + AppSec remote activation

### PR DESCRIPTION
**Pull request by simon-id**
### What does this PR do?
Adds the new Remote Config client, that subscribes to products and receive configuration update for those products.
It also includes the first use case: AppSec 1 click activation.

It repurposes the `Scheduler` class as it was not used anywhere else.

It also adds the BigInt global in the ESLint conf.

### Motivation
This removes a lot of friction to enable AppSec on services, and will enable many more use cases in the future.

### CODE REVIEW

File: packages/dd-trace/src/proxy.js
Comment (simon-id) [inline packages/dd-trace/src/proxy.js:61]: TODO: rename to ASM_FEATURES

File: packages/dd-trace/src/proxy.js
Comment (simon-id) [inline packages/dd-trace/src/proxy.js:67]: TODO: protect against double enable

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [inline packages/dd-trace/src/remote_config/index.js:100]: TODO: this can be removed once #2387 is merged

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [inline packages/dd-trace/src/remote_config/index.js:27]: // TODO: do we need hostname/port here ?

**Comment by simon-id** [comment]
Integration tests are failing with one span being called `http.request`instead of `web.request` but I rebased twice, I thought this was fixed ?

**Comment by juan-fernandez** [comment]
> Integration tests are failing with one span being called `http.request`instead of `web.request` but I rebased twice, I thought this was fixed ?

one of the tests might need an extra `DD_TRACE_TELEMETRY_ENABLED` 

I think it was only set for a test (the one that was failing then): https://github.com/DataDog/dd-trace-js/pull/2422/files#diff-aef7d5c9ebb5bb277297013f0d492dde1b78b2913a1a7e2dae02adead1aebc15R148

**Comment by simon-id** [comment]
@juan-fernandez it's the same test failing :thinking: and it has the `DD_TRACE_TELEMETRY_ENABLED: 'false'`

**Comment by simon-id** [comment]
Wait my PR is sending an outgoing http request at startup, could that fail the test the same way ? @juan-fernandez

**Comment by juan-fernandez** [comment]
> Wait my PR is sending an outgoing http request at startup, could that fail the test the same way ? @juan-fernandez

yes it could

**Comment by simon-id** [comment]
If I disable the remote config requests, the test passes. Instead of disabling remote config (not currently possible), could we just fix the damn test instead ? :sweat_smile:

File: packages/dd-trace/src/proxy.js
Comment (rochdev) [inline packages/dd-trace/src/proxy.js:62]: Let's always use curly braces except for short-circuits to make the code clearer.

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [inline packages/dd-trace/src/remote_config/index.js:1]: Should this be kept in `appsec` for now? We currently have multiple remote config functionalities and should probably have a cross-team discussion before committing to a final approach.

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [inline packages/dd-trace/src/remote_config/index.js:14]: Why is that the case? Should we ensure that this cannot be done instead of documenting to not do it?

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [inline packages/dd-trace/src/remote_config/index.js:20]: We no longer use a scheduler for tracing and instead just schedule as needed after the task has run once. It would be more consistent to do this here, and it would also make it easier to migrate to long-polling later on.

Definitely not a blocker and it's fine either way.

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [reply packages/dd-trace/src/remote_config/index.js:27]: Do we?

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [inline packages/dd-trace/src/remote_config/index.js:31]: Is this format documented publicly anywhere? At worst the agent code would do, but we should probably have a reference somewhere to know what should be included, especially as things are added.

File: packages/dd-trace/src/util.js
Comment (rochdev) [inline packages/dd-trace/src/util.js:13]: Why is this function needed?

File: packages/dd-trace/src/remote_config/scheduler.js
Comment (rochdev) [inline packages/dd-trace/src/remote_config/scheduler.js:27]: Using an interval can be dangerous for HTTP requests if the timeout is higher than the interval. It's safer to instead schedule the next call after the previous call is done.

File: packages/dd-trace/test/proxy.spec.js
Comment (rochdev) [inline packages/dd-trace/test/proxy.spec.js:166]:
```suggestion
      it('should enable appsec when explicitly configured to true', () => {
```

File: packages/dd-trace/test/proxy.spec.js
Comment (rochdev) [inline packages/dd-trace/test/proxy.spec.js:176]:
```suggestion
      it('should not enable appsec but listen to remote config when appsec is not explicitly configured', () => {
```

File: packages/dd-trace/test/remote_config/index.spec.js
Comment (rochdev) [inline packages/dd-trace/test/remote_config/index.spec.js:1]: Any way to write these tests as integration tests and use the tracer directly instead of adding more unit tests?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/remote_config/index.js:27]: Telemetry did include the three props for it to work with UDS so i'll do that too.

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/remote_config/index.js:1]: You mean just moving the folder under the appsec folder ? It doesn't change anything for me. Who said this would be the "final approach" we can always iterate on the interface very easily.

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/remote_config/index.js:14]: Because the agent and backend expect to only have 1 state per tracer, and doing otherwise would cause some unexpected message delivery inconsistencies I'd say.
Since tracer.init() already has double call protection, I'd say we don't really need to enforce it. The only reason this comment is here is because I know I have bad memory.

File: packages/dd-trace/test/proxy.spec.js
Comment (simon-id) [inline packages/dd-trace/test/proxy.spec.js:224]:
```suggestion
      it('should not enable appsec when explicitly configured to false', () => {
```

File: packages/dd-trace/test/remote_config/index.spec.js
Comment (simon-id) [reply packages/dd-trace/test/remote_config/index.spec.js:1]: As we discussed privately, I'm totally up for it but I think we have to initiate this push with examples and tests we can copy from, then sync to all start the effort together.

File: packages/dd-trace/src/remote_config/scheduler.js
Comment (simon-id) [reply packages/dd-trace/src/remote_config/scheduler.js:27]: I know and that's why I added a path that checks if the request is still running. If the callback hasn't been called when the new interval kicks in, it just skips this iteration.
I've never liked to do setTimeouts in a loop but I don't remember why. It just feels weird to me. You think I should implement it with setTimeouts instead ?

File: packages/dd-trace/src/util.js
Comment (simon-id) [reply packages/dd-trace/src/util.js:13]: Because the config for AppSec is not just a boolean now, it can be in three different states;
- true
- false
- {not set}

It is that way because we only want to enable remote config when it is not set (the local config takes precedence).
Maybe you have a better idea to implement this three state check ?

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/remote_config/index.js:31]: There is the protobuff definition file [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/proto/datadog/remoteconfig/remoteconfig.proto), but it doesn't have explanations on the fields.

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/remote_config/index.js:20]: I don't think I understand how that could work here. Traces are pushing info to the backend, here we are polling info from the backend, we don't have a queue of tasks, the backend does.

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [reply packages/dd-trace/src/remote_config/index.js:1]: We already have an existing approach for CI visibility, so we should iterate on that if anything. Otherwise I'd say yes let's move under appsec.

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [reply packages/dd-trace/src/remote_config/index.js:20]: I guess the timeout is smaller than the poll interval anyway, so it doesn't really matter and can be addressed later if needed.

File: packages/dd-trace/src/remote_config/scheduler.js
Comment (rochdev) [reply packages/dd-trace/src/remote_config/scheduler.js:27]: `setInterval` is better suited when something should happen every X amount of time exactly. In our case, it's better that a new timer starts after the operation is finished to avoid any overlaps. Basically we wait for X time after each iteration instead of from the previous schedule. It's also worth noting that for long-polling this would basically be required since a new poll should always happen right after the previous one has finished.

File: packages/dd-trace/src/remote_config/index.js
Comment (rochdev) [reply packages/dd-trace/src/remote_config/index.js:31]: Better than nothing, at least it documents the available fields.

File: packages/dd-trace/src/util.js
Comment (rochdev) [reply packages/dd-trace/src/util.js:13]: Do you mean there are 3 final states? What are they? Can remote config be partially enabled?

Comment (simon-id) [reply packages/dd-trace/src/util.js:13]: This is what I'm trying to implement:
image
As you can see the local conf (`DD_APPSEC_ENABLED`) can be in three different states: `true`, `false`, and "not set".
We only activate remote appsec activation if `DD_APPSEC_ENABLED` is "not set".
This is why I need `isTrueOrFalse()`, I need to differentiate between truthy, falsy, and undefined-ish.

File: integration-tests/startup.spec.js
Comment (simon-id) [inline integration-tests/startup.spec.js:148]: Rolling back the dirty fix from #2422 here.

**Review by bengl — DISMISSED** [review]

File: integration-tests/startup/index.js
Comment (simon-id) [reply integration-tests/startup/index.js:19]: @bengl You sure we should merge that ? I don't even know why it works

File: packages/dd-trace/src/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/remote_config/index.js]: Yes my previous implementation was only using buffers and views but the spec asked for an arbitrary amount of tags, so BigInt it is :(

File: integration-tests/startup/index.js
Comment (simon-id) [reply integration-tests/startup/index.js:19]: My guess on what is happening here: 
This stealthy require integration test fails without this line because it expects outbounds requests from the tracer itself to not be traced.
But by deleting the require cache of the datadog-core AsyncLocalStorage instance (as shown above) it creates two different instances of AsyncLocalStorage and breaks the code that prevents the tracer from tracing itself.
So by adding this line, I prevent this ALS instance mismatch and the integration test passes.

Happy to show you on zoom if the explanation is too obscure.
cc: @tlhunter

File: .eslintrc.json
Comment (tlhunter) [inline .eslintrc.json:13]: If you were to instead remove this new `globals` section and then add `"es2020": true` to the `env` section does it still run? I think that would be preferable.

File: integration-tests/startup/index.js
Comment (tlhunter) [reply integration-tests/startup/index.js:19]: My intuition is we'll need to find another solution to this.

File: packages/dd-trace/src/appsec/remote_config/capabilities.js
Comment (tlhunter) [inline packages/dd-trace/src/appsec/remote_config/capabilities.js:4]: I'm curious why these small numbers (2, 4, 8) are represented as `BigInt`s?

File: packages/dd-trace/src/appsec/remote_config/index.js
Comment (tlhunter) [inline packages/dd-trace/src/appsec/remote_config/index.js:51]: Maybe do something like this:

```js
const DEFAULT_CAPABILITY = Buffer.from('whatever', 'hex').toString('base64')

// ...

        capabilities: DEFAULT_CAPABILITY // updated by `updateCapabilities()`
```

Otherwise the `'AA=='` string looks very magical!

File: packages/dd-trace/src/appsec/remote_config/index.js
Comment (tlhunter) [inline packages/dd-trace/src/appsec/remote_config/index.js:244]: What does this `employee` part mean?

File: packages/dd-trace/src/appsec/remote_config/capabilities.js
Comment (tlhunter) [reply packages/dd-trace/src/appsec/remote_config/capabilities.js:4]: It looks like it has to do with encoding?

File: integration-tests/startup/index.js
Comment (rochdev) [reply integration-tests/startup/index.js:19]: Doesn't that mean we'd have the same issue when a customer is actually using `stealthy-require`?

File: packages/dd-trace/src/appsec/remote_config/index.js
Comment (rochdev) [inline packages/dd-trace/src/appsec/remote_config/index.js:81]: Can this be fixed instead?

File: packages/dd-trace/src/proxy.js
Comment (rochdev) [inline packages/dd-trace/src/proxy.js:57]: Can this be moved elsewhere? It doesn't really belong in the proxy.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/util.js
Comment (simon-id) [reply packages/dd-trace/src/util.js:13]: Look at this line: https://github.com/DataDog/dd-trace-js/pull/2401/files#diff-dae9ab64bee2d029cc08e9fdb70b3cd9d2f897e9b451d325e51033d457bb2994R54
This undefined value is sourced by "isTrueOrFalse" when the config is neither explicitely true or false, undefined means "not locally set". We can zoom if you still think we can do this otherwise.

File: packages/dd-trace/src/proxy.js
Comment (simon-id) [reply packages/dd-trace/src/proxy.js:57]: You're right, what do you think about moving this to appsec's index.js ? that would mean switching from the "dirty require" that we have right now to prevent any appsec code to be run unless specifically enabled, to a normal top of file require.

File: packages/dd-trace/src/appsec/remote_config/scheduler.js
Comment (simon-id) [reply packages/dd-trace/src/appsec/remote_config/scheduler.js:11]: No other parts of the code was using the scheduler, its usecase was removed in a previous PR so I just repurposed the file basically.

File: packages/dd-trace/src/appsec/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/appsec/remote_config/index.js:81]: What do you want to fix ?

File: packages/dd-trace/src/appsec/remote_config/index.js
Comment (simon-id) [reply packages/dd-trace/src/appsec/remote_config/index.js:244]: From the spec:
>The source of the configuration. Currently we only have datadog for customer-issued configurations and employee for datadog employee-issued configurations.

File: integration-tests/startup/index.js
Comment (simon-id) [reply integration-tests/startup/index.js:19]: well if stealthy-require messes with our datadog-core storage instance, then yeah I guess the tracer could be acting weird, but it's not a problem with this PR, it's a problem that already exists in the tracer.

Comment (simon-id) [reply integration-tests/startup/index.js:19]: Here is the PR to separate the issue

File: packages/dd-trace/src/util.js
Comment (simon-id) [reply packages/dd-trace/src/util.js]: Alright I've changed it to `appsec_enabled && isTrue(appsec_enabled)` and it looks like it's working. I thought I previously tried it but I guess I was wrong

File: packages/dd-trace/src/proxy.js
Comment (Qard) [inline packages/dd-trace/src/proxy.js:63]: Seems to me like the purpose of "action" before this conditional and after are different. Wouldn't it be clearer to split that logic into separate variables?

File: packages/dd-trace/src/proxy.js
Comment (Qard) [reply packages/dd-trace/src/proxy.js:57]: Seems to me more like it should live in the config code itself. 🤔

File: packages/dd-trace/src/proxy.js
Comment (simon-id) [reply packages/dd-trace/src/proxy.js:63]: Yeah I have a bad abit of reusing variables, i'll clear that up

Comment (Qard) [reply packages/dd-trace/src/proxy.js:63]: Reusing variables with different types can sometimes cause deopts. It's best to avoid.

Comment (simon-id) [reply packages/dd-trace/src/proxy.js:63]: Ah alright, thanks for the tip :+1:

File: packages/dd-trace/src/proxy.js
Comment (simon-id) [reply packages/dd-trace/src/proxy.js]: I've moved it not in the config.js file as that would be too akward imo, but it it's own index.js under the remote_config folder. We can always change it later but for now since RC is only used by appsec i think it makes sense.

**Comment by tlhunter** [comment]
@simon-id can you rebase this PR on master?

File: packages/dd-trace/src/config.js
Comment (rochdev) [inline packages/dd-trace/src/config.js:204]: Given the loose equality operator, isn't this the same as before?

Comment (rochdev) [reply packages/dd-trace/src/config.js:204]: Same below.

File: packages/dd-trace/src/config.js
Comment (rochdev) [inline packages/dd-trace/src/config.js:215]: The right side of this will not work properly with this operator. Is it even needed if `appsec.enabled` is a boolean anyways? `isTrue` is meant for strings.

File: packages/dd-trace/src/config.js
Comment (simon-id) [reply packages/dd-trace/src/config.js:204]: the loose operator is to check if it's null or undefined, to be backward compatible with the experimental option, it had nothing to do with the isTrueOrFalse thing from before we talked about ?

File: packages/dd-trace/src/config.js
Comment (simon-id) [reply packages/dd-trace/src/config.js:215]: yeah ok i got confused here, i can just have
```suggestion
      appsec.enabled,
```
but it means any value passed will be evaluated a truthy/falsy

File: packages/dd-trace/src/config.js
Comment (simon-id) [reply packages/dd-trace/src/config.js:217]: Right i got confused again

Comment (simon-id) [inline packages/dd-trace/src/config.js:217]:
```suggestion
      process.env.DD_APPSEC_ENABLED && isTrue(process.env.DD_APPSEC_ENABLED)
```

File: packages/dd-trace/src/config.js
Comment (simon-id) [reply packages/dd-trace/src/config.js:204]: It's different from the previous version because if the value of `options.appsec` is false, I want to use that false value. I only bypass it if it's null or undefined

Comment (rochdev) [reply packages/dd-trace/src/config.js:205]: Ah yeah gotcha. I'm not a fan of loose equality generally speaking and would have preferred a check for a boolean, but  not gonna block on this.

**Review by rochdev — APPROVED** [review]
LGTM, but it looks like integration tests are failing again on `http.request` vs `web.request`.

**Comment by simon-id** [comment]
@rochdev Yes the stealthy-require integ test is failing because you guys told me to move the dirty fix I made out of this PR, so I did: https://github.com/DataDog/dd-trace-js/pull/2537

**Pull request merged**
